servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/IdleTimeoutConnectionFilter.java

```diff
@@ -0,0 +1,242 @@
+                    requests = MIN_VALUE;
+                    return failed(new RetryableClosedChannelException(delegate(), timeoutNs));
+                }
+                return delegate().request(request)
+                        .liftSync(new BeforeFinallyHttpOperator(() -> {
+                            // It's acceptable to use 2 volatile variables instead of a single object state here. Even
+                            // if 2 threads race between updating "lastResponseTime" and "requests", the delay for a new
+                            // timer task will be close to "timeoutNs".
+                            requestsUpdater.updateAndGet(this, prev -> {
+                                if (prev > 1) {
+                                    return prev - 1;
+                                }
+                                assert prev == 1 : "Unexpected requests value: " + prev;
+                                lastResponseTime = nanoTime();
+                                return 0;
+                            });
+                        })).shareContextOnSubscribe();
+            });
+        }
+
+        private void updateIdleTimeout(final long delayNs) {
+            final Cancellable newTask = timeoutExecutor.schedule(this, delayNs, NANOSECONDS);
+            if (!timeoutTaskUpdater.compareAndSet(this, null, newTask)) {
+                assert timeoutTask == CANCELLED : "Unexpected timeoutTask: " + timeoutTask;
+                newTask.cancel();    // Connection was closed, cancel the new task
+            }
+        }
+
+        @Override
+        public void run() {
+            final Cancellable oldTask = timeoutTaskUpdater.getAndSet(this, null);
+            if (oldTask == CANCELLED) {
+                // Connection already closed
+                return;
+            }
+            for (;;) {
+                final long requests = this.requests;
+                if (requests > 0) {
+                    // Reschedule timeout:
+                    updateIdleTimeout(timeoutNs);
+                    return;
+                } else if (requests == 0) {
+                    final long nextDelayNs = timeoutNs - (nanoTime() - lastResponseTime);
+                    if (nextDelayNs <= 0) {
+                        if (requestsUpdater.compareAndSet(this, 0, MIN_VALUE)) {
+                            FilterableStreamingHttpConnection connection = delegate();
+                            LOGGER.debug("Closing connection {} after {} ms of inactivity",
+                                    connection, NANOSECONDS.toMillis(timeoutNs));
+                            connection.closeAsync().subscribe();
+                            return;
+                        }
+                    } else {
+                        updateIdleTimeout(nextDelayNs);
+                        return;
+                    }
+                } else {
+                    // Should never happen. Keep it just in case to prevent infinite loop.
```

Comment:
this may happen unless we atomically prevent overflow.

Reply:
Since this is a connection level filter, it always works after concurrency controller, which will make sure we never have more than `Integer.MAX_VALUE` concurrent requests on a single connection. Technically, the overflow can happen for a reserve connections.

I don't think using `addWithOverflowProtection` is the right choice here as it will lead to a problem of incorrectly detecting when connection is idle. 
For example, if there are `MAX_VALUE + 1` concurrent requests on a connection (for h2 that extra request will fail asap, for h1 it's possible to have quite long pipeline), finishing the first `MAX_VALUE` requests will return the counter back to 0, but there might be still 1 request running. 

WDYT if this filter will start rejecting new requests instead of overflowing? `Integer.MAX_VALUE`  seems a reasonable values users should not exceed. Alternatively, we can switch to `Long` and start rejecting only for long overflow.

Reply:
discussed offline. rejecting to prevent overflow makes sense.